### PR TITLE
Const generics & heapless v0.7

### LINKED ISSUE #90: Consts generics tracking
This issue serves as a reminder of where consts generics can be applied instead of the currrent `TypeNum` implementation.

- `AtatLen` could probably be completely removed in favor of consts generics
- Queues (`ComQueue`, `ResQueue`, `UrcQueue`)
- `BufLen`

All of the above are dependent on #81

### PR DESCRIPTION
**Pull request by MathiasKoch**
- Changes all `typenum` usage to const generics
- Bumps `heapless` to 0.7

Fixes #90

### CODE REVIEW

**Comment by fkohlgrueber** [comment]
I'd like to update to heapless 0.7. Any objections to that? If not, I can prepare a PR.

**Comment by MathiasKoch** [comment]
I have a few commits to this locally.. i got a bit stuck on some associated const, but i think I have figured out a solution.. can push it tomorrow, and tag you for a review?

**Comment by fkohlgrueber** [comment]
Sure, sounds good!

**Comment by MathiasKoch** [comment]
I just pushed my WIP, but i still have some issues.

My problem is around `AtatCmd::CommandLen` and turning that into a const generic, which is now required by heapless v0.7.

Issue is, that `min_const_generics` is not enough to use associated consts in generic consts spots. This will yield 
```
error: generic parameters may not be used in const operations
  --> atat/src/traits.rs:84:37
   |
84 |     fn as_bytes(&self) -> Vec<u8, { Self::CommandLen::USIZE }>;
   |                                     ^^^^^^^^^^^^^^^^^^^^^^^ cannot perform const operation using `Self`
   |
   = note: type parameters may not be used in const expressions
   = help: use `#![feature(const_generics)]` and `#![feature(const_evaluatable_checked)]` to allow generic const expressions
```

I am not sure how to get around this? I can't even see a way around it if we introduce `typenum` for parts of it again?

Any inputs are valuable, as i would love to see it all use const generics!

**Comment by fkohlgrueber** [comment]
Oh, that's tough!

I'm not an expert regarding const generics, but I think this would really need full `const_generics` unfortunately.

I tried to find a workaround and the best thing I could come up with is making the `AtatCmd` generic over its length:

```rust
pub trait AtatCmd<const LEN: usize> {
    type Response: AtatResp;
    type Error: FromStr;

    fn as_bytes(&self) -> Vec<u8, LEN>;
    ...
}
```

This then requires adaptions in other places, e.g.:
```rust
// before
fn send<A: AtatCmd>(&mut self, cmd: &A) ...;
// after
fn send<A: AtatCmd<LEN>, const LEN: usize>(&mut self, cmd: &A) ...;
```

Implementation of `AtatCmd` can still be done only on a single length: 
```rust
pub struct CloseConnection {}

impl AtatCmd<15> for CloseConnection {
    type Response = String<256>;
    type Error = GenericError;

    fn as_bytes(&self) -> Vec<u8, 15> {
        ...
    }
    ...
}
```

I haven't fully tried this / thought this through, but this might work. What do you think?

**Comment by fkohlgrueber** [comment]
I wasn't sure how using this would look like, so I set up a simple example:

```rust
pub trait AtatCmd<const N: usize> {
    fn as_bytes(&self) -> [u8; N];
}

/// Command implementing AtatCmd for a single LEN
struct MyCmd;

impl AtatCmd<15> for MyCmd {
    fn as_bytes(&self) -> [u8; 15] {
        [0; 15]
    }
}

/// Command implementing AtatCmd for two different LENs
struct MyCmd2;

impl AtatCmd<30> for MyCmd2 {
    fn as_bytes(&self) -> [u8; 30] {
        [1; 30]
    }
}

impl AtatCmd<100> for MyCmd2 {
    fn as_bytes(&self) -> [u8; 100] {
        [42; 100]
    }
}

fn send<A: AtatCmd<LEN>, const LEN: usize>(cmd: &A) {
    let bytes = cmd.as_bytes();
    println!("Bytes: {:?}", bytes);
}


fn main() {
    // MyCmd implements AtatCmd only for a single LEN, so send can be called 
    // just as usual
    let cmd = MyCmd;
    send(&cmd);
    
    // MyCmd2 implements AtatCmd for two different LENs, so we need to annotate 
    // the send function call to specify which implementation to use.
    let cmd2 = MyCmd2;
    // simply calling send like this will not compile ("type annotations needed"):
    // send(&cmd2); 
    // turbofish to the rescue:
    send::<_,30>(&cmd2);
    send::<_,100>(&cmd2);
}
```

This looks pretty good to me. Compared to having the length as an associated const, it's possible to implement `AtatCmd` for multiple lengths instead of just one. I'm not sure if there are use-cases for having multiple impls with different lengths for the same command, but it doesn't seem to hurt either. If there's only a single impl, the command type can be used in generic functions (e.g. `send`) just as before. If there are multiple implementations, we need to use turbofish in the call to specify which implementation to call. To me that seems like an acceptable solution until more `const_generics` are stabilized.

**Comment by MathiasKoch** [comment]
I think that could work. 
Let me try to implement it, and push the changes here, so we can have a look in a broader context.

Implementation incoming...

**Comment by MathiasKoch** [comment]
Only place i am a bit in doubt if this could work, is when introducing the helper trait of `AtatLen` while deriving. 

It looks like this 
```rust
pub trait AtatLen {
    const LEN: usize;
}
```

The issue i see is that the `AtatCmd` impl would become
```rust
impl AtatCmd<<Self as AtatLen>::LEN> for MyCmd2 {
    fn as_bytes(&self) -> Vec<u8, <Self as AtatLen>::LEN> {
        ...
    }
}
```

Not sure if that works?

**Comment by MathiasKoch** [comment]
Hmm, got close :/ 

```rust
impl AtatCmd<{ <Self as AtatLen>::LEN }> for MyCmd2 {
    fn as_bytes(&self) -> Vec<u8, { <Self as AtatLen>::LEN }> {
        ...
    }
}
```

With the result `generic 'Self' types are currently not permitted in anonymous constants` :/

**Comment by fkohlgrueber** [comment]
Ah, so this doesn't work for things like `Vec<T, L>` or `String<L>`. I currently have no idea how this could be fixed then.

**Comment by MathiasKoch** [comment]
I think we have worked out a fix ;) Hang tight! :+1:

**Comment by MathiasKoch** [comment]
I just pushed a fix.

Would you give it a review @fkohlgrueber ?

I think the only limitation of this way of doing it is that `#[derive(AtatCmd)]` cannot be used on structs with generics. Instead they have to be manually implemented.

File: atat/src/queues.rs [hunk 1]
```diff
@@ -1,47 +1,46 @@
...
 
 // Consumers
-pub type ComConsumer = Consumer<'static, ComItem, ComCapacity, u8>;
-pub type ResConsumer<BufLen> = Consumer<'static, ResItem<BufLen>, ResCapacity, u8>;
-pub type UrcConsumer<BufLen, UrcCapacity> = Consumer<'static, UrcItem<BufLen>, UrcCapacity, u8>;
+pub type ComConsumer = Consumer<'static, ComItem, { COM_CAPACITY + 1 }>;
+pub type ResConsumer<const BUF_LEN: usize> =
+    Consumer<'static, ResItem<BUF_LEN>, { RES_CAPACITY + 1 }>;
+pub type UrcConsumer<const BUF_LEN: usize, const URC_CAPACITY: usize> =
+    Consumer<'static, UrcItem<BUF_LEN>, URC_CAPACITY>;
```
Comment (fkohlgrueber) [inline atat/src/queues.rs:24]: not sure if this should be `{ URC_CAPACITY + 1}`

File: atat/src/traits.rs [hunk 2]
```diff
@@ -64,15 +63,15 @@ pub trait AtatUrc {
...
-    type CommandLen: ArrayLength<u8>;
+pub trait AtatCmd<const LEN: usize> {
+    // /// The max length of the command.
+    // ///
+    // /// Example: For the command "AT+RST" you would specify
+    // ///
+    // /// ```
+    // /// const COMMAND_LEN: usize = 6;
+    // /// ```
+    // const COMMAND_LEN: usize;
```
Comment (fkohlgrueber) [inline atat/src/traits.rs:74]: delete?

File: atat/src/traits.rs [hunk 3]
```diff
@@ -180,37 +185,24 @@ pub trait AtatClient {
...
         &self,
         resp: Result<&[u8], &InternalError>,
     ) -> Result<Self::Response, Error<Self::Error>> {
-        resp.map(|r| {
-            heapless::String::from_utf8(Vec::from_slice(r).map_err(|_| Error::Parse)?)
-                .map_err(|_| Error::Parse)
-        })
-        .map_err(Error::from)?
+        let str = core::str::from_utf8(resp.map_err(Error::from)?).map_err(|_| Error::Parse)?;
+        Ok(String::from(str))
```
Comment (fkohlgrueber) [inline atat/src/traits.rs:205]: `str` as variable name is confusing (even confuses GitHub's syntax highlighting).

File: serde_at/src/de/mod.rs [hunk 4]
```diff
@@ -45,21 +45,16 @@ pub type Result<T> = core::result::Result<T, Error>;
 /// assert_eq!(incoming, expected);
 /// ```
 #[derive(Debug, Clone, PartialEq)]
-pub struct CharVec<T: heapless::ArrayLength<char>>(pub heapless::Vec<char, T>);
+pub struct CharVec<const T: usize>(pub heapless::Vec<char, T>);
```
Comment (fkohlgrueber) [inline serde_at/src/de/mod.rs:48]: Minor thing: I'd prefer `N` instead of `T` as it's a length and not a type.

File: serde_at/src/ser/mod.rs [hunk 5]
```diff
@@ -426,30 +408,28 @@ where
 }
 
 /// Serializes the given data structure as a string
-pub fn to_string<B, C, T>(
+pub fn to_string<T, const B: usize, const C: usize>(
```
Comment (fkohlgrueber) [inline serde_at/src/ser/mod.rs:411]: Is there a convention to list const generics after generic types? Just curious...

**Review by fkohlgrueber — APPROVED** [review]
Looks good to me, even though I'm not entirely sure I understand the derive macro stuff. I've commented a couple of smaller things.

File: atat/src/queues.rs (hunk 1 above)
Comment (MathiasKoch) [reply atat/src/queues.rs:24]: It should preferrably be, but i can't seem to make it work. 

It comes from the `N - 1` initializer here: https://japaric.github.io/heapless/heapless/spsc/struct.Queue.html

File: serde_at/src/ser/mod.rs (hunk 5 above)
Comment (MathiasKoch) [reply serde_at/src/ser/mod.rs:411]: It's a requirement

**Pull request merged**